File: backend/src/models/Note.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface INote extends Document {
  userId: mongoose.Types.ObjectId;
  title: string;
  content: string;
  tags: string[];
  linkedEntities: mongoose.Types.ObjectId[];
  createdAt: Date;
  updatedAt: Date;
  embeddings?: number[];
}

const noteSchema = new Schema<INote>(
  {
    userId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    title: { type: String, required: true },
    content: { type: String, required: true },
    tags: [{ type: String }],
    linkedEntities: [{ type: Schema.Types.ObjectId, ref: 'Entity' }],
    embeddings: [{ type: Number }],
  },
  { timestamps: true }
);

// Index for faster querying
noteSchema.index({ userId: 1, createdAt: -1 });
noteSchema.index({ tags: 1 });

export const Note = mongoose.model<INote>('Note', noteSchema);

File: backend/models/note.py
"""
Note model with AI embeddings and entity linking
"""

from beanie import Document, PydanticObjectId
from pydantic import Field
from typing import List, Optional
from datetime import datetime


class Note(Document):
    """Note document model with AI capabilities"""
    
    user_id: PydanticObjectId
    title: str
    content: str
    embeddings: List[float] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    linked_people: List[PydanticObjectId] = Field(default_factory=list)
    linked_meetings: List[PydanticObjectId] = Field(default_factory=list)
    entities: dict = Field(default_factory=dict)  # Extracted entities
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "notes"
        indexes = [
            "user_id",
            "created_at",
            [("user_id", 1), ("created_at", -1)],
        ]
    
    class Config:
        json_schema_extra = {
            "example": {
                "title": "Meeting with Sarah",
                "content": "Discussed project timeline and deliverables",
                "tags": ["work", "project"],
                "entities": {
                    "people": ["Sarah"],
                    "topics": ["project timeline", "deliverables"]
                }
            }
        }
